File: src/git.zig
const std = @import("std");

pub const Git = struct {
    allocator: std.mem.Allocator,
    io: std.Io,

    const max_diff_size = 1024 * 1024;

    pub fn init(allocator: std.mem.Allocator, io: std.Io) Git {
        return .{
            .allocator = allocator,
            .io = io,
        };
    }

    pub fn deinit(self: Git) void {
        _ = self;
    }

    pub fn getStagedDiff(self: Git) ![]const u8 {
        return try self.runCaptureChecked(&.{ "git", "diff", "--cached" }, max_diff_size, error.GitDiffFailed);
    }

    pub fn stageAll(self: Git) !void {
        try self.runChecked(&.{ "git", "add", "-A" }, error.GitStageFailed);
    }

    pub fn commit(self: Git, message: []const u8) !void {
        try self.runChecked(&.{ "git", "commit", "-m", message }, error.GitCommitFailed);
    }

    pub fn push(self: Git) !void {
        try self.runChecked(&.{ "git", "push" }, error.GitPushFailed);
    }

    fn runChecked(self: Git, argv: []const []const u8, failure: anyerror) !void {
        const result = try std.process.run(self.allocator, self.io, .{
            .argv = argv,
            .stdout_limit = .limited(1024 * 1024),
            .stderr_limit = .limited(1024 * 1024),
        });
        defer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);

        switch (result.term) {
            .exited => |code| if (code != 0) {
                printCommandOutput(result);
                return failure;
            },
            else => {
                printCommandOutput(result);
                return failure;
            },
        }
    }

    fn runCaptureChecked(self: Git, argv: []const []const u8, max_bytes: usize, failure: anyerror) ![]const u8 {
        const result = try std.process.run(self.allocator, self.io, .{
            .argv = argv,
            .stdout_limit = .limited(max_bytes),
            .stderr_limit = .limited(1024 * 1024),
        });
        errdefer self.allocator.free(result.stdout);
        defer self.allocator.free(result.stderr);

        switch (result.term) {
            .exited => |code| if (code != 0) {
                printCommandOutput(result);
                return failure;
            },
            else => {
                printCommandOutput(result);
                return failure;
            },
        }

        return result.stdout;
    }

    fn printCommandOutput(result: std.process.RunResult) void {
        if (result.stdout.len > 0) std.debug.print("{s}", .{result.stdout});
        if (result.stderr.len > 0) std.debug.print("{s}", .{result.stderr});
    }
};

test "Git - initialization" {
    const testing = std.testing;
    const allocator = testing.allocator;

    const git = Git.init(allocator, std.testing.io);
    defer git.deinit();

    // Validate structural initialization invariants.
    try testing.expect(@TypeOf(git.allocator) == std.mem.Allocator);
}

File: src/git.rs
use std::process::{Command, ExitStatus};

use thiserror::Error;

const MAX_DIFF_BYTES: usize = 1024 * 1024;

#[derive(Debug, Error)]
pub enum GitError {
    #[error("Git command failed: {0}")]
    Command(#[from] std::io::Error),
    #[error("git {command} exited with {status}")]
    Failed {
        command: &'static str,
        status: ExitStatus,
    },
    #[error("the staged diff exceeds the {MAX_DIFF_BYTES} byte limit")]
    DiffTooLarge,
    #[error("Git returned non-UTF-8 diff output")]
    InvalidUtf8(#[from] std::string::FromUtf8Error),
}

pub struct Git;

impl Git {
    pub fn stage_all() -> Result<(), GitError> {
        run("add", ["add", "-A"])
    }

    pub fn staged_diff() -> Result<String, GitError> {
        let output = Command::new("git")
            .args(["diff", "--cached", "--no-ext-diff", "--binary"])
            .output()?;
        if !output.status.success() {
            return Err(GitError::Failed {
                command: "diff",
                status: output.status,
            });
        }
        if output.stdout.len() > MAX_DIFF_BYTES {
            return Err(GitError::DiffTooLarge);
        }
        Ok(String::from_utf8(output.stdout)?)
    }

    pub fn commit(message: &str) -> Result<(), GitError> {
        run("commit", ["commit", "-m", message])
    }

    pub fn push() -> Result<(), GitError> {
        run("push", ["push"])
    }
}

fn run<'a>(command: &'static str, args: impl IntoIterator<Item = &'a str>) -> Result<(), GitError> {
    let status = Command::new("git").args(args).status()?;
    if status.success() {
        Ok(())
    } else {
        Err(GitError::Failed { command, status })
    }
}
